### ISSUE TITLE
模型训练很少的几个Epoch后就验证集上的loss就不再降

### ISSUE THREAD

**Issue by YEN-GitHub** [shenweichen/DeepCTR#89]
**Describe the question(问题描述)**
我用的模型是[DeepFM](https://github.com/shenweichen/DeepCTR/blob/master/deepctr/models/deepfm.py),跑的数据集是完整的criteo. 参数基本用的代码默认的，一些修改过的参数为：embedding_size=10, dnn_hidden_units=(400, 400, 400)，optimizater=tf.keras.optimizers.Adam(lr=0.0001), 我加入了ealry_stop机制，但发现验证集loss基本几个epoch后就不下降了

`
if __name__ == "__main__":
    data = pd.read_csv('./data/criteo/data.txt')
    # data = pd.read_csv('./data/criteo/data_sample.txt')
    sparse_features = ['C' + str(i) for i in range(1, 27)]
    dense_features = ['I' + str(i) for i in range(1, 14)]

    data[sparse_features] = data[sparse_features].fillna('-1', )
    data[dense_features] = data[dense_features].fillna(0, )
    target = ['label']

    print("# 1.Label Encoding for sparse features,and do simple Transformation for dense features")
    for feat in sparse_features:
        lbe = LabelEncoder()
        data[feat] = lbe.fit_transform(data[feat])
    mms = MinMaxScaler(feature_range=(0, 1))
    data[dense_features] = mms.fit_transform(data[dense_features])

    print("# 2.count #unique features for each sparse field,and record dense feature field name")
    sparse_feature_list = [SingleFeat(feat, data[feat].nunique())
                           for feat in sparse_features]
    dense_feature_list = [SingleFeat(feat, 0,)
                          for feat in dense_features]

    print("# 3.generate input data for model")
    train, test = train_test_split(data, test_size=0.2)
    train_model_input = [train[feat.name].values for feat in sparse_feature_list] + \
                        [train[feat.name].values for feat in dense_feature_list]
    test_model_input  = [test[feat.name].values for feat in sparse_feature_list] + \
                        [test[feat.name].values for feat in dense_feature_list]

    print("# 4.Define Model,train,predict and evaluate")
    model = DeepFM({"sparse": sparse_feature_list, "dense": dense_feature_list},
                    embedding_size=10, dnn_hidden_units=(400, 400, 400), task='binary')

    model.compile(tf.keras.optimizers.Adam(lr=0.0001), "binary_crossentropy",
                  metrics=['binary_crossentropy'], )

    early_stopping = EarlyStopping(monitor='val_loss', patience=2, verbose=2)
    history = model.fit(train_model_input, train[target].values,
                        batch_size=4096, epochs=64, verbose=2, validation_split=0.1,\
                        callbacks=[early_stopping])

    pred_ans = model.predict(test_model_input, batch_size=4096)
    print("test LogLoss", round(log_loss(test[target].values, pred_ans), 4))
    print("test AUC", round(roc_auc_score(test[target].values, pred_ans), 4))
`

训练过程：
`
D:\pycharmCommunity\Anaconda3\python.exe E:/RecSys/Lib/DeepCTR-tf/runs/run_deepfm_criteo.py
# 1.Label Encoding for sparse features,and do simple Transformation for dense features
# 2.count #unique features for each sparse field,and record dense feature field name
# 3.generate input data for model
# 4.Define Model,train,predict and evaluate
WARNING:tensorflow:From E:\RecSys\Lib\DeepCTR-tf\deepctr\layers\interaction.py:427: calling reduce_sum (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
WARNING:tensorflow:From D:\pycharmCommunity\Anaconda3\lib\site-packages\tensorflow\python\ops\init_ops.py:1184: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with distribution=normal is deprecated and will be removed in a future version.
Instructions for updating:
`normal` is a deprecated alias for `truncated_normal`
Train on 33005243 samples, validate on 3667250 samples
Epoch 1/64
2019-06-22 05:19:31.443070: I T:\src\github\tensorflow\tensorflow\core\platform\cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2019-06-22 05:20:07.726690: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1405] Found device 0 with properties:
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.683
pciBusID: 0000:01:00.0
totalMemory: 11.00GiB freeMemory: 9.11GiB
2019-06-22 05:20:11.448823: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1484] Adding visible gpu devices: 0
2019-06-22 05:21:23.341445: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-06-22 05:21:23.341662: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:971]      0
2019-06-22 05:21:23.341805: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:984] 0:   N
2019-06-22 05:21:23.378332: I T:\src\github\tensorflow\tensorflow\core\common_runtime\gpu\gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 8807 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
 - 552s - loss: 0.4658 - binary_crossentropy: 0.4640 - val_loss: 0.4556 - val_binary_crossentropy: 0.4528
Epoch 2/64
 - 276s - loss: 0.4544 - binary_crossentropy: 0.4509 - val_loss: 0.4529 - val_binary_crossentropy: 0.4490
Epoch 3/64
 - 277s - loss: 0.4511 - binary_crossentropy: 0.4466 - val_loss: 0.4523 - val_binary_crossentropy: 0.4477
Epoch 4/64
 - 273s - loss: 0.4491 - binary_crossentropy: 0.4440 - val_loss: 0.4520 - val_binary_crossentropy: 0.4467
Epoch 5/64
 - 269s - loss: 0.4477 - binary_crossentropy: 0.4420 - val_loss: 0.4525 - val_binary_crossentropy: 0.4467
Epoch 6/64
 - 275s - loss: 0.4467 - binary_crossentropy: 0.4405 - val_loss: 0.4528 - val_binary_crossentropy: 0.4466
Epoch 7/64
 - 275s - loss: 0.4458 - binary_crossentropy: 0.4392 - val_loss: 0.4527 - val_binary_crossentropy: 0.4461
Epoch 00007: early stopping
test LogLoss 0.4456
test AUC 0.8059

Process finished with exit code 0
`

**Operating environment(运行环境):**
 - python 3.6.5 :: Anaconda, Inc.
 - tensorflow version  1.10.0 
 - deepctr version v0.4.1

**Comment by shenweichen** [comment]
说明再继续训练就要过拟合了~

**Comment by YEN-GitHub** [comment]
> 说明再继续训练就要过拟合了~

是的  验证集是反应的这样的结论    可我想问一下：criteo的数据里大概四千万，deepfm参数也很多，只用训练4,5个epoch就能训练好吗

**Comment by shenweichen** [comment]
可以的，一般1个epoch就足够了

**Comment by YEN-GitHub** [comment]
> 可以的，一般1个epoch就足够了

好的 多谢

**Comment by xxllp** [comment]
数据量越大训练次数肯定越少啊

**Comment by guotong1988** [comment]
> 可以的，一般1个epoch就足够了

学习了
